nfqueue.c

```diff
@@ -14,84 +14,152 @@
 */
 
 #include <stdio.h>
-#include <stdint.h>
+#include <stdlib.h>
 
 #include <arpa/inet.h>
-#include <libnetfilter_queue/libnetfilter_queue.h>
+
+#include <libmnl/libmnl.h>
 #include <linux/netfilter.h>
+#include <linux/netfilter/nfnetlink.h>
+
+#include <linux/netfilter/nfnetlink_queue.h>
+
+#include <libnetfilter_queue/libnetfilter_queue.h>
 
 #include "nfqueue.h"
 
 extern int debug;
 
-static int recv_callback(struct nfq_q_handle *qh, struct nfgenmsg *nfmsg,
-			 struct nfq_data *nfad, void *data);
+static int recv_callback(const struct nlmsghdr *nlh, void *data);
+
+static struct mnl_socket *nl;
 
 int nfqueue_receive(uint16_t queue_num, int (*callback)())
 {
-	int fd;
-	ssize_t len;
-	char buf[1024];
-	struct nfq_handle *h;
-	struct nfq_q_handle *qh;
-
-	h = nfq_open();
-	if (!h) {
-		fprintf(stderr, "Failed opening nfq\n");
-		return 1;
-	}
+	size_t sizeof_buf = 0xFF + MNL_SOCKET_BUFFER_SIZE / 2;
+	uint16_t portid;
+	char *buf;
+	struct nlmsghdr *nlh;
 
-	nfq_unbind_pf(h, AF_INET);
-	if (nfq_bind_pf(h, AF_INET) < 0) {
-		fprintf(stderr, "Failed to bind nfnetlink\n");
-		return 1;
+	if (debug)
+		puts("Setting up netlink socket");
+
+	// Create socket
+	nl = mnl_socket_open(NETLINK_NETFILTER);
+	if (nl == NULL) {
+		fprintf(stderr, "mnl_socket_open() failed\n");
+		return EXIT_FAILURE;
 	}
 
-	qh = nfq_create_queue(h, queue_num, recv_callback, callback);
-	if (!qh) {
-		fprintf(stderr, "Failed to bind socket to queue\n");
-		return 1;
+	if (mnl_socket_bind(nl, 0, MNL_SOCKET_AUTOPID) < 0) {
+		fprintf(stderr, "mnl_socket_bind() failed\n");
+		return EXIT_FAILURE;
 	}
 
-	nfq_set_queue_flags(qh, NFQA_CFG_F_FAIL_OPEN, NFQA_CFG_F_FAIL_OPEN);
+	portid = mnl_socket_get_portid(nl);
+
+	buf = malloc(sizeof_buf);
```

Comment:
I don't see that the allocated memory is free'd anywhere. Why not stick to the original stack-allocated buffer?